To support lightsleep for Heltec V3/v4 repeater to achieve 9mA power consumption

IoTThinks opened the pull request:
Hi all,
This PR is to allow Heltec v3 and v4 repeaters to save power down from 47mA to 9mA most of the time.
One 2500mAh can help to last 2 weeks for Heltec v3 and v4.
- To use light sleep to achieve 9mA
- No latency due to quicker wakeup
- First boot: To wake up for 30s to send an advert at 18th second.
- Light sleep: To wake up when receiving a LoRa packet and light sleep again after 5s.
- Periodically wakeup: To wake up every 30 minutes to do some periodically tasks and and lightsleep again after 5s.
- Do not sleep when WiFi is not off likely due to OTA via WiFi.
- To reboot to out of OTA and back to powersaving mode.

First advert and periodically adverts (if set) are working.
OTA via WiFi works too.

The code should be applicable to all ESP32-S3 MCUs.
We can move the code upstream. However, at this stage, I put at Heltec v3 and v4 first.
Thanks a lot.

SaschaKt commented:
Works like a charm. Should be standard on all esp repeaters or possible to turn on/off via commandline

IoTThinks commented:
Let me monitor.
Usually, I keep power it on / off so it will be back to 2024. And I can always sync as it is 2024.

Let me sync clock and let it run.

By right, the clock of esp32 is not accurate. It will drift...

SaschaKt commented:
> @SaschaKt You check time using clock CLI, right?
> 
> ![Screenshot_20251120_160731_MeshCore.jpg](https://private-user-images.githubusercontent.com/29994971/516746088-0eb9b2cd-bcd7-4759-bd37-da9275800329.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM2MzQwMTgsIm5iZiI6MTc2MzYzMzcxOCwicGF0aCI6Ii8yOTk5NDk3MS81MTY3NDYwODgtMGViOWIyY2QtYmNkNy00NzU5LWJkMzctZGE5Mjc1ODAwMzI5LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIwVDEwMTUxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5NGUzMDNhMDc3OWVhYjhmZTk0NzFiNDI2YThlNDcwNmUyZjQzNWUyMjdkNmJjMmZkOGM1YjMxYTk2ZWUzOWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.g1AgPdHp909YA7pDPtTy9_F2Cp61AHefZkJcPFKxrks)

I look at the Request Status page and in cli. The time goes ahead and cannot be synced. Only resetbutton on v3 sets it to 2024, the v4 is not in the near so I can't hit the reset button

IoTThinks commented:
@SaschaKt My Heltec v3's clock works fine for 2 hours. Let me monitor a day and see.
BTW, the clock in CLI is in UTC not in timezone.

How far is the drift in one day should I expect?

mikecarper commented:
https://analyzer.letsme.sh/node/7E7662676F7F0850A8A355BAAFBFC1EB7B4174C340442D7D7161C9474A2C9400

This is cougar; probably the most busy node in all of MeshCore with roughly 110 neighbors. The granularity is 10s; there are savings to be found even here 
<img width="3740" height="538" alt="image" src="https://github.com/user-attachments/assets/69dfb7e6-0240-4983-bc44-3184c9cae124" />

cparen commented:
I just want to add my +1 to this change.

I've got my own private branch BLE companion that does lightsleep anytime the Bluetooth isn't connected, periodically waking up to allow reconnect. This would make it easier to keep my private branch in sync with main.

bosco688 commented:
Works great on Heltec v4

IoTThinks commented on variants/heltec_v3/HeltecV3Board.h:
```diff
@@ -76,6 +77,23 @@ class HeltecV3Board : public ESP32Board {
+
+  void sleep() override {
+    if (WiFi.getMode() == WIFI_MODE_NULL) { // WiFi is off ~ No active OTA, can go to sleep
+      enterLightSleep(1800); // To wake up every 30 minutes or when receiving a LoRa packet
```
The 30-minute is to wake up every 30 minutes to do periodically jobs like adverts.

Other times, it will go to sleep every 5s.

IoTThinks replied:
Ok let me make the sleep duration as parameters for sleep method.

IoTThinks commented:
@mikecarper The dashboard looks awesome.

Looks like an ISP operator even this is P2P network.

IoTThinks commented:
> > @SaschaKt You now have 2 minutes instead of 30s to setup a Repeater via Serial USB.
> > When we refresh the Repeater Setup page, the esp32 will be resetted and we have another 2 minutes to do the settings again.
> > After that, it will sleep after 5s.
> 
> Hi Kevin. I noticed with the v3 and v4 that the clock go ahead, the longer the running time. A normal reboot via cli doesn't set the clock to the 2024 Standard time so it is not possible to sync the clock (because time can not be synched if it is correct or ahead). Can you modify the CLI "clock Sync" so that it is always syncing?
> 
> It is interesting why the clock goes ahead, do you have noticed?

@SaschaKt 
1. During light sleep, ESP32 will use slower and less accurate clock to save power.
That is why the clock is not as accurate as it is when in full operation.
And Heltec v3/v4 does not have external clock which is more accurate during sleeps.

Actually, why we need the clock anyway? In embedded system like LoRaWAN, we generally dont have clock at dumb devices like sensors and repeaters.

As long as the clock does not move ...backward, then it should be fine.

2. We can change code to sync the clock if the time is in advance.
However, this is usually not recommended due to potential file corruption or packet timestamp mismatch...
So I skip this approach.

3. Yes, it is interesting as MeshCore does not reset the time when we send reboot in MeshCore app.
From the code, yes, it seems it will keep the time during software reboot.
IMO, yes, it should reset everything the same as when we hit RESET button.

<img width="1042" height="300" alt="image" src="https://github.com/user-attachments/assets/f057918e-eb24-45f0-9504-ec09d49ca0e6" />

So I may leave the timing issue as it is now.

aqua commented:
> What are the odds of getting this working with the seeed esp33 node? https://www.seeedstudio.com/Wio-SX1262-with-XIAO-ESP32S3-p-5982.html

It works almost as-is on the seeed esp32 -- only problem on my test build is that it doesn't wake properly on LoRa packets.  Probably just incorrectly setting the wake conditions, but I haven't tried to troubleshoot it yet.

IoTThinks commented:
Let me move the code upstream to apply to all ESP32 based board to avoid duplicated code.
To enable, we can set a flag to enable the powersaving feature in platformio.ini of each board.

aqua commented:
> @aqua Let me enable power saving mode and compile for XIAO.
> 
> Currently, I set the flag to enable for Heltec v3/v4 only.
> 
> Did you make changes and compiled for Xiao?

Yes, I just replicated sleep() and enterLightSleep() in XiaoS3WIOBoard and flashed to a Xiao S3 unit on a power meter.  The arguments to esp_sleep_enable_ext1_wakeup() will probably need to be customized for some boards (e.g. with user buttons) but that might be the only variant override required and the default behavior could go into ESP32Board for most of them.

IoTThinks commented:
@aqua I have moved up the code for sleep() and enterLightSleep() from Heltec v3/v4 to ESP32Board.h.
And I added   -D POWERSAVING_MODE=1 for Xiao Wio too.

For now, I believe all ESP32 should share the same sleep() and enterLightSleep() to avoid code duplication.
I believe most of the ESP32 based repeaters do not use buttons. PROG button does not have any effect too. 
They only use...RST button. 
We can add it in later phase or in variant if anyone wants.

For your testing, you can compile yourself or use my pre-compiled bin files at 
https://github.com/IoTThinks/EasySkyMesh/tree/main/firmware

It has the CPU temperature at Telemetry > Channel 1 too.
<img height="384" alt="image" src="https://github.com/user-attachments/assets/0f165833-621d-4e1b-a429-81d6d878df23" />

mavericm1 commented:
Been noticing some strange behavior.  I may need to get my sdr to double check but when a message is sent from my companion it isn't immediately repeated and it seems to wait until another message is sent.   

This happens after upgrading the v4 to this pr but a meshcore app update happened at the same time so it could be either causing this behavior

mavericm1 commented:
> @mavericm1 Yes, please help to test with SDR. Let me check with my SDR during weekend too.
> 
> Btw if so then MeshCore app will show the retries, right?

confirmed on the SDR that it isn't repeating how it should 

https://limewire.com/d/qkXQP#RUy3lm9qNb

zjs81 commented on src/helpers/ESP32Board.h:
```diff
@@ -42,6 +43,23 @@ class ESP32Board : public mesh::MainBoard {
   #endif
   }
 
+  void enterLightSleep (uint32_t secs) {
```
A better solution would be to use the GPIO wakeup source 

GPIO Wakeup (Light-sleep Only)[](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html#gpio-wakeup-light-sleep-only)
In addition to EXT0 and EXT1 wakeup sources described above, one more method of wakeup from external inputs is available in Light-sleep mode. With this wakeup source, each pin can be individually configured to trigger wakeup on high or low level using [gpio_wakeup_enable()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/gpio.html#_CPPv418gpio_wakeup_enable10gpio_num_t15gpio_int_type_t) function. Unlike EXT0 and EXT1 wakeup sources, which can only be used with RTC IOs, this wakeup source can be used with any IO (RTC or digital).

Should be like this 


```
void enterLightSleep(uint32_t s){
  pinMode(P_LORA_DIO_1,INPUT);
  esp_sleep_disable_wakeup_source(ESP_SLEEP_WAKEUP_ALL);
  gpio_wakeup_enable((gpio_num_t)P_LORA_DIO_1,GPIO_INTR_HIGH_LEVEL);
  esp_sleep_enable_gpio_wakeup();
  if(s>0) esp_sleep_enable_timer_wakeup((uint64_t)s*1000000ULL);
  esp_light_sleep_start();
}

void sleep(uint32_t s) override {
  if(WiFi.getMode()==WIFI_MODE_NULL) enterLightSleep(s);
}
```


Then at the top you gotta add. 

```
#include "esp_sleep.h"
#include "driver/gpio.h"
```

zjs81 replied:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/sleep_modes.html#gpio-wakeup-light-sleep-only

IoTThinks replied:
@zjs81 Let me use and test your GPIO wakeup to maximize the compability for any ESP32-based board.
Thanks a lot. 👯

IoTThinks replied:
@zjs81 I tested.
The wakeup via GPIO will not work for DIO1 of sx1262.

Dio1 of sx1262 will have short rising HIGH.
GPIO wake up may trigger many times and crash esp32 when trying to wakeup.
Let me see if any work around.

In my experience, I always use ext0 and ext1 for dio1 of sx1262.

I may change to use ext0 instead of ext1 as some boards are esp32-c3.
Esp32-c3 only has ext0 and no ext1.

SaschaKt commented:
> > @mavericm1 Yes, please help to test with SDR. Let me check with my SDR during weekend too.
> > Btw if so then MeshCore app will show the retries, right?
> 
> confirmed on the SDR that it isn't repeating how it should
> 
> https://limewire.com/d/qkXQP#RUy3lm9qNb

This behaviour I have noticed since a longer time and bevor Iotthinks modificated Firmware. Sometimes, mostly after a longer period of inactivity or when I write to another group it happens. Try please with the official latest release, I am sure this behaviour is generally a problem

kinkywaterhose commented:
> > > > > > @mavericm1 Yes, please help to test with SDR. Let me check with my SDR during weekend too.
> > > > > > Btw if so then MeshCore app will show the retries, right?
> > > > > 
> > > > > 
> > > > > confirmed on the SDR that it isn't repeating how it should
> > > > > https://limewire.com/d/qkXQP#RUy3lm9qNb
> > > > 
> > > > 
> > > > This behaviour I have noticed since a longer time and bevor Iotthinks modificated Firmware. Sometimes, mostly after a longer period of inactivity or when I write to another group it happens. Try please with the official latest release, I am sure this behaviour is generally a problem
> > > 
> > > 
> > > I run another repeater with same settings and official build vs the v4 lightsleep build and it doesn't have this problem.
> > 
> > 
> > Do you have more repeaters in range. With 3 repeaters I have much problems, I've tried all, set txdelay and so on. With one v4 as repeater it goes now very good. Maybe try flashing..my V4 has the light sleep on. The only thing is that the clock goes every day more ahead with v3/v4 and lightsleep hack
> 
> yes i have more than 1 repeater within range. This isn't an issue in the main branch and further talking about it in this PR isn't helping anything if you think its a bug in the main branch then report it as an issue. If you want to fix your repeaters that are all in range then do `set int.thresh 14` on them

I ran into this issue trying to create power saving features, and worked around it by checking if there is work to be done in Dispatcher. If there is work to be done, don't sleep until the work is cleared up. (it required a bit of work, but corrected this issue since the radio sleeps too quick for the packet to be sent out)

IoTThinks commented:
Interesting, let me check over SDR.

We have tested to send from a companion to another companion via 2 hops with no issue.

Usually, this should be very straight forward with light sleep for esp32.

mavericm1 commented:
> Interesting, let me check over SDR.
> 
> We have tested to send from a companion to another companion via 2 hops with no issue.
> 
> Usually, this should be very straight forward with light sleep for esp32.

Try using set int.thresh 14 on the repeater running this maybe that is the trigger that is causing it to act that way in combination to light sleep

IoTThinks commented:
> Xiao S3 doesn't wake on packet with that build -- not surprising since it's the same wake bitmask just in a different place. Might be a few days before I have time to fiddle with it, so anyone with differently-distributed free time, feel free.

@aqua Xiao S3 Wio uses GPIO39 as DIO1 and this is not a RTC pin.
So the wakeup via ext1 (best practice) does not work.

Actually, this is not a good hardware design. If going to **deep**sleep, this board can not wake up when a LoRa paket comes.
For light sleep, let me see how to workaround for Xiao S3.

IoTThinks commented:
To ensure the maximum compability.
- I removed the   -D POWERSAVING_MODE=1 for XIAO WIO S3 since its DIO1 (GPIO39) is not a RTC GPIO. This is a safety feature. We need to manually add this in the platform.ini to enable this feature. So default is for Heltec v3 and Heltec v4. Later we may add to all ESP32-S3 with RTC GPIO for DIO1.
- I added if defined to support ESP32-S3 only. Some ESP32-C3, C6 may not have EXT0 and EXT1. Sleep will skip for these boards.
- I added validation for RTC GPIO for DIO1. If the DIO1 is not RTC, sleep will skip for these boards.

I have compiled for ALL targets and all target can compile.
- Heltec_ct62_repeater_bridge_espnow has unrelated compilation issue. Not in scope. Skip.
- Heltec_Wireless_Tracker_repeater_bridge_espnow and Heltec_Wireless_Tracker_room_server refer to outdated  #include <helpers/HeltecV3Board.h>. Not in scope. Skip.

IoTThinks commented:
@mavericm1 I dont get what you mean.

You mean the repeater repeats twice or the repeater needs to wait the next slot to repeat?

mavericm1 commented:
> > Message gets stuck in the queue and doesn't go out until another event wakes it up. If I'm reading it correctly
> 
> @mikecarper I guess so too. So let me fix the potential issue: "Delay the sleep until no more dispatcher task". I suspect due to high traffic and high inteference, the repeater delays the repeat and can not finish repeating in 5s.

Yes what mike said is what i was describing here.  https://github.com/meshcore-dev/MeshCore/pull/1107#issuecomment-3597246193

IoTThinks commented:
@mavericm1 @mikecarper Your testing and feedback are well appreciated.

How about this? 
I will check if there is anything to send. 
If no, sleep. 
Else, it will awake for 5 more seconds.
<img height="384" alt="image" src="https://github.com/user-attachments/assets/cd864bed-5063-4fdc-a574-3a34305da2f8" />

Some friends here suggest me to implement this feature dynamically instead of hardcode parameter.
So in CLI of repeater, we can do "set powersaving on" to turn on this feature. 
The default is well, off.

What do you think?

mavericm1 commented:
> @mavericm1 @mikecarper Your testing and feedback are well appreciated.
> 
> How about this? I will check if there is anything to send. If no, sleep. Else, it will awake for 5 more seconds.
> 
> I will test and see how it reacts first. Then let you test in the busy area.
> <img alt="image" height="384" src="https://private-user-images.githubusercontent.com/29994971/522464829-5317d843-3e83-4471-83e9-d3f76ccb2781.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ4ODc4MjgsIm5iZiI6MTc2NDg4NzUyOCwicGF0aCI6Ii8yOTk5NDk3MS81MjI0NjQ4MjktNTMxN2Q4NDMtM2U4My00NDcxLTgzZTktZDNmNzZjY2IyNzgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA0VDIyMzIwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI3NTFiNGMwOGRkZGVmZjVlMzczN2EwMjkwOWZhNzQ3NGI2YmQzNDgzYmQ1YjFhYmI2NDNkNDcxYWQ1ODc1MGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AeYVEsUf-CxG1QyZfK-2Y9kvZFhUNVSkRNlUmvDL1xc">
> 
> Some friends here suggest me to implement this feature dynamically instead of hardcoded parameter. So in CLI of repeater, we can do "set powersaving on" to turn on this feature. The default is well, off.
> 
> What do you think?

Yes i think that is what makes the most sense that the dispatcher doesn't allow sleep until queue is empty

IoTThinks commented:
@mavericm1 I have tested. Looks OK. The getOutboundCount() is retrieved from Dispatcher.
Note the delay(100) is for debug log during testing only. This is to print the message properly before sleep.

In the photo, we can see when getOutboundCount=1, the sleep will be postponed from second 44 to second 49.
<img height="384" alt="image" src="https://github.com/user-attachments/assets/aaacdd0d-6923-4276-afde-ae0572140376" />

I put the pre-compiled bin file here for your testing.
They have the printout in the console in cause you want to check.
https://github.com/IoTThinks/EasySkyMesh/tree/main/firmware/Testing

Please note the logic:
- First boot: To wake up for 2 minutes to send an advert at 18th second and allow Repeater Setup via UART cable.
- Light sleep: To wake up when receiving a LoRa packet and light sleep again after 5s. To postpone 5 more seconds if there is pending outbound send.
- Periodically wakeup: To wake up every 30 minutes to do some periodically tasks and and try to lightsleep again after 5s.

Please help to test and feedback.
Thanks a lot.

IoTThinks commented:
So good.
I'm in a correct track.

hpux735 commented:
I wasn't able to reproduce any savings with a v4.  Here's a screen shot of power analysis before:
<img width="2258" height="1682" alt="image" src="https://github.com/user-attachments/assets/bd5ef1c2-5bdd-4acf-a018-a2a2a16972d4" />

And after applying this MR to a local branch.
<img width="2258" height="1738" alt="ppk-20251210T222711" src="https://github.com/user-attachments/assets/aef2fe29-e916-4cb7-b7c4-667583a94cd5" />

The quiescent current is 59mA in both cases.  Do you have any suggestions?  Or, perhaps, a bin file for the v4 that I could test to ensure that it's being compiled as you expect?

hpux735 commented:
I tried with the bin you linked to and got exactly the same results.

SaschaKt commented:
> I tried with the bin you linked to and got exactly the same results.

Have you tried another v4? Is the v4 maybe charging the battery if you measure over the Usb-C port (red led must be off)? It takes 2 minutes for the first sleep after switching on or rebooting

IoTThinks commented on examples/simple_repeater/MyMesh.cpp:
```diff
@@ -1101,3 +1101,8 @@ void MyMesh::loop() {
   last_millis = now;
 }
+
+// To get the current pending outbound packets at Dispatcher
```
Ok, let me rename it to hasPendingWork().

IoTThinks commented on examples/simple_repeater/main.cpp:
```diff
@@ -117,4 +122,16 @@ void loop() {
   rtc_clock.tick();
+
+#ifdef POWERSAVING_MODE
+  if (millis() - lastActive > nextSleepinSecs * 1000) {
```
Ok, let me take a look at millisHasNowPassed()
Thanks a lot.

ViezeVingertjes commented:
> We can convert all ESP32 boards from another mesh to become MeshCore repeaters at 9mA. Much lower than any NRF52 boards with BLE at 14+ mAh.

Woah, 17 µA is not 14+ mAh haha... that's a whole different ballpark. That 17 µA is from the specs, as my meters wont even measure that low i dont know the exact number, but way below 1mA when it's on (in sleep with wake upon rx) until it gets an packet and repeats it.
